src/WriteClient.ts

```diff
@@ -0,0 +1,1323 @@
+	// 					typeof assetOrID === "string" ? assetOrID : assetOrID.id,
+	// 				),
+	// 			},
+	// 			params,
+	// 		}),
+	// 	)
+	// }
+
+	/**
+	 * Fetches a foreign asset from a URL.
+	 *
+	 * @param url - The URL of the asset to fetch.
+	 * @param params - Additional fetch parameters.
+	 *
+	 * @returns A file representing the fetched asset.
+	 */
+	private async fetchForeignAsset(
+		url: string,
+		params: FetchParams = {},
+	): Promise<File> {
+		const requestInit: RequestInitLike = {
+			...this.fetchOptions,
+			...params.fetchOptions,
+			headers: {
+				...this.fetchOptions?.headers,
+				...params.fetchOptions?.headers,
+			},
+			signal:
+				params.fetchOptions?.signal ||
+				params.signal ||
+				this.fetchOptions?.signal,
+		}
+
+		const res = await this.fetchFn(url, requestInit)
+
+		if (!res.ok) {
+			throw new PrismicError("Could not fetch foreign asset", url, undefined)
+		}
+
+		const blob = await res.blob()
+
+		return new File([blob], "", {
+			type: res.headers.get("content-type") || undefined,
+		})
+	}
+
+	/**
+	 * {@link resolveAssetTagIDs} rate limiter.
+	 */
+	private _resolveAssetTagIDsLimit = pLimit()
+
+	/**
+	 * Resolves asset tag IDs from tag names or IDs.
+	 *
+	 * @param tagNamesOrIDs - An array of tag names or IDs to resolve.
+	 * @param params - Whether or not missing tags should be created and
+	 *   additional fetch parameters.
+	 *
+	 * @returns An array of resolved tag IDs.
+	 */
+	private async resolveAssetTagIDs(
+		tagNamesOrIDs: string[] = [],
+		{ createTags, ...params }: { createTags?: boolean } & FetchParams = {},
+	): Promise<string[]> {
+		return this._resolveAssetTagIDsLimit(async () => {
+			const existingTags = await this.getAssetTags(params)
+			const existingTagMap: Record<string, AssetTag> = {}
+			for (const tag of existingTags) {
+				existingTagMap[tag.name] = tag
+			}
+
+			const resolvedTagIDs = []
+			for (const tagNameOrID of tagNamesOrIDs) {
+				if (isAssetTagID(tagNameOrID)) {
+					resolvedTagIDs.push(tagNameOrID)
+					continue
+				}
+
+				// Tag does not exists yet, we create it if `createTags` is set
+				if (!existingTagMap[tagNameOrID] && createTags) {
+					existingTagMap[tagNameOrID] = await this.createAssetTag(
+						tagNameOrID,
+						params,
+					)
+				}
+
+				// Add tag if found
+				if (existingTagMap[tagNameOrID]) {
+					resolvedTagIDs.push(existingTagMap[tagNameOrID].id)
+				}
+			}
+
+			return resolvedTagIDs
+		})
+	}
+
+	/**
+	 * Creates a tag in the asset API.
+	 *
+	 * @remarks
+	 * Tags should be at least 3 characters long and 20 characters at most.
+	 *
+	 * @param name - The name of the tag to create.
+	 * @param params - Additional fetch parameters.
+	 *
+	 * @returns The created tag.
+	 */
+	private async createAssetTag(
+		name: string,
+		params?: FetchParams,
+	): Promise<AssetTag> {
+		if (name.length < 3 || name.length > 20) {
+			throw new PrismicError(
+				"Asset tag name must be at least 3 characters long and 20 characters at most",
```

Comment:
Agreed, let's remove everything like that from the write client.

However, I would keep it in the `Migration` class as I prefer an early throw in it rather than a late throw in `WriteClient#migrate()`